azblob/sas_service.go

```diff
@@ -4,17 +4,16 @@ import (
 	"bytes"
 	"fmt"
 	"strings"
-	"time"
 )
 
 // BlobSASSignatureValues is used to generate a Shared Access Signature (SAS) for an Azure Storage container or blob.
 // For more information, see https://docs.microsoft.com/rest/api/storageservices/constructing-a-service-sas
 type BlobSASSignatureValues struct {
 	Version            string      `param:"sv"`  // If not specified, this defaults to SASVersion
 	Protocol           SASProtocol `param:"spr"` // See the SASProtocol* constants
-	StartTime          time.Time   `param:"st"`  // Not specified if IsZero
-	ExpiryTime         time.Time   `param:"se"`  // Not specified if IsZero
-	SnapshotTime       time.Time
+	StartTime          string      `param:"st"`
+	ExpiryTime         string      `param:"se"`
+	SnapshotTime       string
 	Permissions        string  `param:"sp"` // Create by initializing a ContainerSASPermissions or BlobSASPermissions and then call String()
```

Comment:
I am opposed to swapping time.Time out for strings here.

Why? Because when someone wants to create their SAS token and sign it here, they won't know what format to put in.

Personally, I'd prefer we have a pointer to a string for time format here. If it's nil, we should prefer our default time formatting for our SAS token. Then, at lower levels, we can either store the time.Time and the format OR we can store the string. I feel more comfortable with the former, how about you, @JohnRusk and @zezha-msft

Reply:
The reason for using Time is because these are times and code can easily take NOW add an hour (or whatever) to set the expiration time. Can the resolution problem be fixed by just changing the time format string: azblob.SASTimeFormat

Reply:
I would say no. When stringifying the SAS token back out, we'd lose information as we currently do if we did that.

Reply:
That's correct!  The problem is that the signature is generated with a specific string value, and if the SDK rewrites the expiryTime, then the service will not match the signature and fail the request.

To be clear, this affects existing SASs, not a SAS that the SDK itself creates from a given key.

Reply:
Can we get an example of a SAS time string that fails to convert to a Time due to resolution loss.
And, isn't this happening because of the format string? Why can we not change the format string to that resolution is maintained? Does Time not support the required resolution?

Reply:
Changing the format string will fix my scenario, at the expense of breaking signatures that were generated with expirytimes without high precision, because now the problem is in reverse: times without high-precision information will be written out with it zero'd out (instead of none at all).

I'll get you an example broken URL in a bit.

Reply:
Here's another question? Why are you trying to round trip this? When I designed this, I expected most people will use the SDK to create SAS tokens. A lot fewer people might want to parse a SAS just to get the permissions, expiry time, etc. in order to some local processing. 

I never thought anyone would try to round trip the SAS since there really isn't any need to if you already have the original SAS. I'd hate to see the Time support go away in order to support a rare scenario. 

So, I'd like to understand your scenario more to see if it could be a more useful scenario that requires more attention.

Reply:
I don't think time support should go away, I just think we should keep track of the format to help round tripping. AzCopy for instance, cannot handle these "high precision" times in SAS tokens, due to the inability to round trip here.

Reply:
What is the round trip scenario? Why does AzCopy need to round trip SAS tokens?

Reply:
A URL with SAS token gets parsed by the blob SDK when attempting to enumerate, and the SAS token becomes invalid when attempting to make a request to the service, because the time format has changed.

Reply:
For another instance, the SAS token is separated from the URL with the blob SDK by clearing the SAS token values in a BlobURLParts struct, and then restringified down the line.

Reply:
This still isn't clear to me. Do you mean the customer passes a URL with a container or account SAS as a command-line argument. The SDK parses this using BlobURLParts? which has a SAS field with the parsed SAS and then AzCopy for some reason uses this SAS field to generate a new SAS and therefore the signature doesn't match?

If I got this right, then why do this? I guess you want the the SAS parameters parsed out of the URL for you easily, right? If so, why not just add a private string field to SASQueryParameters and have it hold the original SAS query parameters. If any file dis changed, invalid this field (set it to ""). If no field is changed, then have the function that gets a SAS token string return the field (if not "") and have it do what it does today if the field is "".

This way, nothing break; minimal code change, round-tripping now works. Times ALSO continue to work.

Reply:
Yes, the intent was easy removal of the SAS token in that case.

I  don't feel like what you suggested (the private string field to the SAS query structure) is really a fix to the underlying problem here, more than it is sweeping it under the rug.

Reply:
There is still a loss of information going on in that case. The time itself does not have to be a string, but we should be aware of what time format succeeded when parsing the URL, and we should be aware of what time format the user wants in the generation structures, especially if there's a similar bug in another SDK or tool.

Reply:
It's not about SAS token removal because you want to re-apply it, right? I still can't say I full understand the whole flow and what's motivating this. I don't know why you think my proposal doesn't solve the problem or that it sweeps it under the rug. I'm trying to do minimal code change, no breaking changes, address the issue and keep the other features.

Reply:
URL with SAS:
https://asbg169.blob.core.windows.net/example/hello.txt?sv=2015-04-05&sr=c&se=2020-04-14T17%3A00%3A00.0000000Z&sp=rl&sig=INgLizrwmX83P5J13gm03qZjCjAt5d5Xuqg%2BTrdSeqI%3D

In a browser, it works great.  AzCopy chokes, because somewhere along the way the SAS gets rewritten before it heads out to the network (note the error text listing an expiry time that is NOT what was passed in with the URL).

```
c:\...\azcopy_windows_amd64_10.4.0> ./azcopy.exe copy "https://asbg169.blob.core.windows.net/example/hello.txt?sv=2015-04-05&sr=c&se=2020-04-14T17%3A00%3A00.0000000Z&sp=rl&sig=INgLizrwmX83P5J13gm03qZjCjAt5d5Xuqg%2BTrdSeqI%3D" "hello.txt" --recursive
INFO: Scanning...
INFO: Any empty folders will not be processed, because source and/or destination doesn't have full folder support

failed to perform copy command due to error: cannot start job due to error: cannot list blobs. Failed with error -> github.com/Azure/azure-storage-blob-go/azblob.newStorageError, /home/vsts/go/pkg/mod/github.com/!azure/azure-storage-blob-go@v0.7.0/azblob/zc_storage_error.go:42
===== RESPONSE ERROR (ServiceCode=AuthenticationFailed) =====
Description=Server failed to authenticate the request. Make sure the value of Authorization header is formed correctly including the signature.
RequestId:aaed7a13-c01e-0086-8084-0f5354000000
Time:2020-04-10T22:04:55.1065374Z, Details:
   AuthenticationErrorDetail: Signature did not match. String to sign used was rl

2020-04-14T17:00:00Z
/blob/asbg169/example



2015-04-05





   Code: AuthenticationFailed
   GET https://asbg169.blob.core.windows.net/example?comp=list&include=metadata&prefix=hello.txt%2F&restype=container&se=2020-04-14t17%3A00%3A00z&sig=-REDACTED-&sp=rl&sr=c&sv=2015-04-05&timeout=901
   User-Agent: [AzCopy/10.4.0 Azure-Storage/0.7 (go1.13; Windows_NT)]
   X-Ms-Client-Request-Id: [32777584-f87a-4757-6a49-deb8f8c2e1ff]
   X-Ms-Version: [2019-02-02]
   --------------------------------------------------------------------------------
   RESPONSE Status: 403 Server failed to authenticate the request. Make sure the value of Authorization header is formed correctly including the signature.
   Content-Length: [489]
   Content-Type: [application/xml]
   Date: [Fri, 10 Apr 2020 22:04:54 GMT]
   Server: [Microsoft-HTTPAPI/2.0]
   X-Ms-Error-Code: [AuthenticationFailed]
   X-Ms-Request-Id: [aaed7a13-c01e-0086-8084-0f5354000000]
```

Reply:
(there is nothing sensitive in this blob, and the SAS will expire Tuesday @ Midnight Pacific time, lest anyone worry)

Reply:
What I'm suggesting isn't a breaking change-- Changing the time variables to strings is. Keeping them as times, and adding an additional variable to track the format actually used is not.

I say that your solution here sweeps it under the rug because, rather than actually tracking the time format, it just keeps the whole SAS token in addition to the parameters, and then regurgitates it when the user asks for the parameters to be stringified, rather than re-stringifying properly.

Reply:
> URL with SAS:
> [https://asbg169.blob.core.windows.net/example/hello.txt?sv=2015-04-05&sr=c&se=2020-04-14T17%3A00%3A00.0000000Z&sp=rl&sig=[REDACTED]](https://asbg169.blob.core.windows.net/example/hello.txt?sv=2015-04-05&sr=c&se=2020-04-14T17%3A00%3A00.0000000Z&sp=rl&sig=%5BREDACTED%5D)
> 
> In a browser, it works great. AzCopy chokes, because somewhere along the way the SAS gets rewritten before it heads out to the network (note the error text listing an expiry time that is NOT what was passed in with the URL).
> 
> ```
> c:\...\azcopy_windows_amd64_10.4.0> ./azcopy.exe copy "https://asbg169.blob.core.windows.net/example/hello.txt?sv=2015-04-05&sr=c&se=2020-04-14T17%3A00%3A00.0000000Z&sp=rl&sig=INgLizrwmX83P5J13gm03qZjCjAt5d5Xuqg%2BTrdSeqI%3D" "hello.txt" --recursive
> INFO: Scanning...
> INFO: Any empty folders will not be processed, because source and/or destination doesn't have full folder support
> 
> failed to perform copy command due to error: cannot start job due to error: cannot list blobs. Failed with error -> github.com/Azure/azure-storage-blob-go/azblob.newStorageError, /home/vsts/go/pkg/mod/github.com/!azure/azure-storage-blob-go@v0.7.0/azblob/zc_storage_error.go:42
> ===== RESPONSE ERROR (ServiceCode=AuthenticationFailed) =====
> Description=Server failed to authenticate the request. Make sure the value of Authorization header is formed correctly including the signature.
> RequestId:aaed7a13-c01e-0086-8084-0f5354000000
> Time:2020-04-10T22:04:55.1065374Z, Details:
>    AuthenticationErrorDetail: Signature did not match. String to sign used was rl
> 
> 2020-04-14T17:00:00Z
> /blob/asbg169/example
> 
> 
> 
> 2015-04-05
> 
> 
> 
> 
> 
>    Code: AuthenticationFailed
>    GET https://asbg169.blob.core.windows.net/example?comp=list&include=metadata&prefix=hello.txt%2F&restype=container&se=2020-04-14t17%3A00%3A00z&sig=-REDACTED-&sp=rl&sr=c&sv=2015-04-05&timeout=901
>    User-Agent: [AzCopy/10.4.0 Azure-Storage/0.7 (go1.13; Windows_NT)]
>    X-Ms-Client-Request-Id: [32777584-f87a-4757-6a49-deb8f8c2e1ff]
>    X-Ms-Version: [2019-02-02]
>    --------------------------------------------------------------------------------
>    RESPONSE Status: 403 Server failed to authenticate the request. Make sure the value of Authorization header is formed correctly including the signature.
>    Content-Length: [489]
>    Content-Type: [application/xml]
>    Date: [Fri, 10 Apr 2020 22:04:54 GMT]
>    Server: [Microsoft-HTTPAPI/2.0]
>    X-Ms-Error-Code: [AuthenticationFailed]
>    X-Ms-Request-Id: [aaed7a13-c01e-0086-8084-0f5354000000]
> ```

I censored it out still because anyone who gets ahold of the SAS token could make a simple query hundreds of thousands of times and cost you money.

Reply:
I like to live dangerously (and wanted to provide a short-lived repro), but hopefully what remains in the paste is sufficient for understanding of the issue.  Thanks for being a cooler-head for me.

To try to start tying off a long-thread, I started out thinking there wasn't a big need to keep the `Time` around, but after sleeping on it a bit agree that it would be nice to keep.  I'm going to look a little closer at the File SDK change I called out earlier, and just port that around if it's sufficient for this issue, and will fix and then port if it needs additional work (it doesn't contain the high-precision time format, though I don't know if it needs it).  I still don't like that it tries to do any parsing of the string, but I'm fine with that being a me-thing. :smile: